src/Interpreters/InterserverCredentials.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <Poco/Util/AbstractConfiguration.h>
+#include <Common/Exception.h>
+#include <common/logger_useful.h>
+
+namespace DB
+{
+/// InterserverCredentials holds credentials for server (store) and client
+/// credentials (current_*). The container is constructed through `make` and a
+/// shared_ptr is captured inside Context.
+class BaseInterserverCredentials
+{
+public:
+    BaseInterserverCredentials(std::string current_user_, std::string current_password_)
+        : current_user(current_user_), current_password(current_password_)
+            { }
+
+    virtual ~BaseInterserverCredentials() { }
+
+    /// isValidUser implements authentication for InterserverIOHandler
+    virtual std::pair<std::string, bool> isValidUser(std::pair<std::string, std::string> pair) = 0;
+
+    std::string getUser() { return current_user; }
+
+    std::string getPassword() { return current_password; }
+
+
+protected:
+    std::string current_user;
+    std::string current_password;
+};
+
+
+/// NullInterserverCredentials are used when authentication is not configured
+class NullInterserverCredentials : public virtual BaseInterserverCredentials
+{
+public:
+    NullInterserverCredentials(const NullInterserverCredentials &) = delete;
+    NullInterserverCredentials() : BaseInterserverCredentials("", "") { }
+
+    ~NullInterserverCredentials() override { }
+
+    static std::shared_ptr<NullInterserverCredentials> make() { return std::make_shared<NullInterserverCredentials>(); }
+
+    std::pair<std::string, bool> isValidUser(std::pair<std::string, std::string> pair) override
+    {
+        std::ignore = pair;
+        return {"", true};
+    }
+};
+
+
+/// SimpleInterserverCredentials implements authentication using a Store, which
+/// is configured, e.g.
+///    <interserver_http_credentials>
+///        <user>admin</user>
+///        <password>222</password>
+///        <!-- To support mix of un/authenticated clients -->
+///        <!-- <allow_empty>true</allow_empty> -->
+///        <users>
+///            <!-- Allow authentication using previous passwords during rotation -->
+///            <admin>111</admin>
+///        </users>
+///    </interserver_http_credentials>
+class SimpleInterserverCredentials : public virtual BaseInterserverCredentials
+{
+public:
+    using Store = std::vector<std::pair<std::string, std::string>>;
```

Comment:
Maybe `std::unordered_map` will be better? (Will not require iteration through all of the users, but not sure that it worth it, since it should have too much elements, but at least it will make interface cleaner)

Reply:
Can do, but I think it's the same?

Reply:
The usage will became simpler, right now it iterates over vector (instead of using `find()`)

Reply:
Use case was only to have 1 extra user in store (previous credentials). Using a map would allow the admin to use a credential per server. Sgtm

Reply:
Actually a map doesn't work because more than one password (old, new) should work for the same user

Reply:
Ok, then it can be multimap, or at least just replace the looping over store with the `std::find`

Reply:
I replaced with a smap<pair<string, string>, bool>. I'm sure it's not what you had in mind but it's using std::find :)